File: EngineUAB/Run/Videogame/Data/Scripts/Game/Player/player_run_state.lua
class 'CPlayerRunState' (CState)
	function CPlayerRunState:__init() 
		CState.__init(self)
		self.action_2_input = core:get_action_to_input()
	end

	function CPlayerRunState:OnEnter(_CCharacter)
		core:get_debug_gui_manager().debug_render:set_state_name("Run")
	end
	
	function CPlayerRunState:Execute(_CCharacter)
		local l_d = 0.0
		local l_yaw = 0.0
		local l_dir = Vect3f(0.0, 0.0, 0.0)
		local l_move_player = false
		local enemy_detected = nil
		
		--Miramos si hay un target a poder fijarse
		if not _CCharacter.is_target_fixed then
			enemy_detected = _CCharacter:detect_enemy()
		end
		
		if not _CCharacter.locked then
			--El jugador ataca
			if self.action_2_input:do_action('AttackPlayer') then
				if _CCharacter.is_target_fixed then
					_CCharacter.logic_fsm:change_state(_CCharacter.target_attack)
					_CCharacter.graphic_fsm:change_state(_CCharacter.animated_attack)
				else
					_CCharacter.logic_fsm:change_state(_CCharacter.attack)
					_CCharacter.graphic_fsm:change_state(_CCharacter.animated_attack)
				end
			end
			
			--Actualización del yaw a partir del movimiento del mouse
			l_d = self.action_2_input:do_action_mouse('YawPlayer')
			_CCharacter.yaw = _CCharacter.yaw + l_d
			_CCharacter.yaw = angle_filter(_CCharacter.yaw)
			l_yaw = _CCharacter.yaw
		
			--Mira los controles de movimiento
			if self.action_2_input:do_action('MovePlayerUp') then --El player se mueve hacia adelante
				if self.action_2_input:do_action('MovePlayerLeft') then
					l_yaw = l_yaw + math.pi/4
				elseif self.action_2_input:do_action('MovePlayerRight') then
					l_yaw = l_yaw - math.pi/4
				end
				l_dir = Vect3f(math.cos(l_yaw), 0.0, math.sin(l_yaw))
				l_move_player = true
				
			elseif self.action_2_input:do_action('MovePlayerDown') then --El player se mueve hacia atrás
				l_yaw = l_yaw - math.pi
				if self.action_2_input:do_action('MovePlayerLeft') then
					l_yaw = l_yaw - math.pi/4
				elseif self.action_2_input:do_action('MovePlayerRight') then
					l_yaw = l_yaw + math.pi/4
				end
				l_dir = Vect3f(math.cos(l_yaw), 0.0, math.sin(l_yaw))
				l_move_player = true
				
			elseif self.action_2_input:do_action('MovePlayerLeft') then --El player se mueve hacia la izquierda
				l_yaw = l_yaw + math.pi/2
				l_dir = Vect3f(math.cos(l_yaw), 0.0, math.sin(l_yaw))
				l_move_player = true
				
			elseif self.action_2_input:do_action('MovePlayerRight') then --El player se mueve hacia la derecha
				l_yaw = l_yaw - math.pi/2
				l_dir = Vect3f(math.cos(l_yaw), 0.0, math.sin(l_yaw))
				l_move_player = true
			end
		end
		
		--Le aplica la velocidad al movimiento
		l_dir = l_dir * 10.0 * _CCharacter.elapsed_time
		
		--Mueve el controller físico
		_CCharacter.physic_controller:move(l_dir, _CCharacter.elapsed_time)
		
		if l_move_player then
			--Pone de forma correcta los ángulos
			_CCharacter.physic_controller.yaw = l_yaw
			_CCharacter.animated_model.yaw = -(l_yaw * 180.0 / math.pi) + 90.0
			
			--Mira si se hace un salto
			if self.action_2_input:do_action('PlayerJump') then
				_CCharacter.logic_fsm:change_state(_CCharacter.jump)
				_CCharacter.graphic_fsm:change_state(_CCharacter.animated_jump)
			end
		else
			--Cambia de estado a idle
			_CCharacter.logic_fsm:change_state(_CCharacter.idle)
			_CCharacter.graphic_fsm:change_state(_CCharacter.animated_idle)
		end
	end
	
	function CPlayerRunState:OnExit(_CCharacter)
	end
	
	function CPlayerRunState:OnMessage(_CCharacter, _Msg)
		if ( _Msg.msg == msg_attack ) then
			local l_Enemy = get_game_process():get_character_manager():get_enemy_by_id(_Msg.sender)
			local l_received_pain = math.random( ( l_Enemy.properties.strong/ 2 ), l_Enemy.properties.strong) 
			local l_received_pain_to_hit = l_Enemy.properties.strong * 0.95		-- Busco el 95% de la fuerza total
			print_logger(1, "l_received_pain_to_hit----- :"..l_received_pain_to_hit)
			print_logger(1, "l_received_pain----- :"..l_received_pain)
			if ( l_received_pain >= l_received_pain_to_hit  ) then
				_CCharacter.logic_fsm:change_state(_CCharacter.hit)
				_CCharacter.graphic_fsm:change_state(_CCharacter.animated_hit)
			end
			print_logger(1, "PLAYER LIFE :".._CCharacter.properties.life)
			return true
		end
		return false
	end
	
	function CPlayerRunState:__Finalize()
	end

File: trunk/EngineUAB/Code/Videogame/Characters/Player/LogicStates/PlayerRunState.cpp
#include "PlayerRunState.h"
#include "StatesMachine\Telegram.h"
#include "RenderableObjects\AnimatedModel\AnimatedInstanceModel.h"
#include "DebugInfo\DebugRender.h"
#include "DebugGUIManager.h"
#include "Characters\Player\Player.h"
#include "Utils\Random.h"
#include "ActionToInput.h"
#include "GameProcess.h"
#include "EngineProcess.h"
#include "Helpers\MathHelper.h"
#include "PhysicController.h"
#include "Core.h"
#include "Base.h"

#include "Steering Behaviors\SteeringEntity.h"
#include "Characters\StatesDefs.h"
#include "Logger\Logger.h"

#if defined _DEBUG
#include "Memory\MemLeaks.h"
#endif

CPlayerRunState::CPlayerRunState( CCharacter * _pCharacter, const std::string &_Name )
	: CState(_pCharacter, _Name)
	, m_fRotationVelocity(10.f)
	, m_fMaxVelocityMovement(10.f)
	, m_fCurrentVelocityMovement(0.f)
	, m_fAccelerationStart(30.f)
	, m_fAccelerationEnd(65.f)
	, m_bStartState(true)
	, m_bEndState(false)
	, m_LastDirection(v3fZERO)
{

}

CPlayerRunState::~CPlayerRunState()
{

}

void CPlayerRunState::OnEnter( CCharacter* _pCharacter )
{
#if defined _DEBUG
	if( CORE->IsDebugMode() )
	{
		CORE->GetDebugGUIManager()->GetDebugRender()->SetStateName("Run");
	}
#endif

	m_LastDirection				= v3fZERO;
	m_fCurrentVelocityMovement	= 0.f;
	m_bStartState				= true;
	m_bEndState					= false;
}

void CPlayerRunState::Execute( CCharacter* _pCharacter, float _fElapsedTime )
{
	CCharacter *l_pEnemy		= NULL;
	CPlayer *l_pPlayer			= static_cast<CPlayer*>( _pCharacter );
	CActionToInput *l_pInput	= CORE->GetActionToInput();

	//Miramos si hay un target a poder fijarse
	if( !l_pPlayer->IsTargetFixed() )
	{
		l_pEnemy = l_pPlayer->DetectEnemy();
	}

	if( !_pCharacter->GetLocked() )
	{
		//El jugador ataca
		if( l_pInput->DoAction("AttackPlayer") )
		{
			if( l_pPlayer->IsTargetFixed() )
			{
				_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("targetattack1") );
				_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animattack1") );
			}
			else
			{
				_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("attack1") );
				_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animattack1") );
			}
		}

		//Actualización del yaw a partir del movimiento del mouse
		float l_fDelta	= l_pInput->DoActionMouse("YawPlayer");
		float l_fYaw	= _pCharacter->GetYaw() - l_fDelta;
		l_fYaw			= Helper::AngleFilter( l_fYaw );
		_pCharacter->SetYaw( l_fYaw );

		//Mira los controles de movimiento
		bool l_bMovePlayer	= false;
		Vect3f l_Dir		= v3fZERO;

		if( l_pInput->DoAction("MovePlayerUp") )
		{
			if( l_pInput->DoAction("MovePlayerLeft") )
			{
				l_fYaw += FLOAT_PI_VALUE / 4.f;
			}
			else if( l_pInput->DoAction("MovePlayerRight") )
			{
				l_fYaw -= FLOAT_PI_VALUE / 4.f;
			}

			l_bMovePlayer = true;
		}
		else if( l_pInput->DoAction("MovePlayerDown") )
		{
			l_fYaw -= FLOAT_PI_VALUE;

			if( l_pInput->DoAction("MovePlayerLeft") )
			{
				l_fYaw -= FLOAT_PI_VALUE / 4.f;
			}
			else if( l_pInput->DoAction("MovePlayerRight") )
			{
				l_fYaw += FLOAT_PI_VALUE / 4.f;
			}

			l_bMovePlayer = true;
		}
		else if( l_pInput->DoAction("MovePlayerLeft") )
		{
			l_fYaw += FLOAT_PI_VALUE / 2.f;
			l_bMovePlayer = true;
		}
		else if( l_pInput->DoAction("MovePlayerRight") )
		{
			l_fYaw -= FLOAT_PI_VALUE / 2.f;
			l_bMovePlayer = true;
		}


		if( l_bMovePlayer )
		{
			l_Dir = Vect3f( mathUtils::Cos<float>(l_fYaw), 0.f, mathUtils::Sin<float>(l_fYaw) );
			m_LastDirection = l_Dir;

			if( m_bEndState )
			{
				m_bStartState = true;
				m_bEndState = false;
			}
		}
		else
		{
			l_Dir = m_LastDirection;
			m_bEndState = true;
		}


		//La aplica la velocidad al movimiento
		if( m_bStartState )
		{
			m_fCurrentVelocityMovement = m_fCurrentVelocityMovement + m_fAccelerationStart * _fElapsedTime;
			if( m_fCurrentVelocityMovement >= m_fMaxVelocityMovement )
			{
				m_fCurrentVelocityMovement = m_fMaxVelocityMovement;
				m_bStartState = false;
			}
		}
		else if( m_bEndState )
		{
			m_fCurrentVelocityMovement = m_fCurrentVelocityMovement - m_fAccelerationEnd * _fElapsedTime;
			if( m_fCurrentVelocityMovement <= 0.f )
			{
				m_fCurrentVelocityMovement = 0.f;
				m_bEndState = false;
			}
		}
		
		
		l_Dir = l_Dir * m_fCurrentVelocityMovement * _fElapsedTime;

		//Mueve el controller físico
		_pCharacter->GetController()->Move( l_Dir, _fElapsedTime );

		if( l_bMovePlayer )
		{
			//Pone de forma correcta los ángulos
			_pCharacter->GetController()->SetYaw( l_fYaw );
			_pCharacter->GetAnimatedModel()->SetYaw( -mathUtils::Rad2Deg(l_fYaw) + 90.f );

			//Mira si se hace un salto
			if( l_pInput->DoAction("PlayerJump") )
			{
				_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("jump") );
				_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animjump") );
			}
		}
		else if( !l_bMovePlayer && !m_bEndState )
		{
			_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("idle") );
			_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animidle") );
		}
	}
}

void CPlayerRunState::OnExit( CCharacter* _pCharacter )
{
}

bool CPlayerRunState::OnMessage( CCharacter* _pCharacter, const STelegram& _Message )
{
	if( _Message.Msg == Msg_Attack )
	{
		CRandom	l_Randomize;

		CCharacter *l_pEnemy	= static_cast<CGameProcess*>(CORE->GetProcess())->GetCharactersManager()->GetCharacterById(_Message.Sender);
		float l_fReceivedPain	= l_Randomize.getRandFloat( (float)(l_pEnemy->GetProperties()->GetStrong() / 2), (float)l_pEnemy->GetProperties()->GetStrong());
		float l_fPainToHit		= l_pEnemy->GetProperties()->GetStrong() * 0.95f;

		if( l_fReceivedPain >= l_fPainToHit )
		{
			static_cast<CPlayer*>(_pCharacter)->HitToPlayer();
		}

		return true;
	}
	else if( _Message.Msg == Msg_Push )
	{
		CCharacter *l_pEnemy	= static_cast<CGameProcess*>(CORE->GetProcess())->GetCharactersManager()->GetCharacterById(_Message.Sender);
		
		sDireccion * l_Info = (struct sDireccion *) _Message.ExtraInfo;
		_pCharacter->MoveTo2(l_Info->Direccion * 1.2f, l_Info->ElapsedTime);
		_pCharacter->GetSteeringEntity()->SetVelocity(Vect3f(0,0,0));
		LOGGER->AddNewLog(ELL_INFORMATION, "CPlayerRunState::OnMessage -> PUSHED!!");
		return true;
	}

	return false;
}